aiapy/psf/psf.py

```diff
@@ -123,70 +110,70 @@ def calculate_psf(channel: u.angstrom, *, use_preflightcore=False, diffraction_o
-    if HAS_CUPY and use_gpu:
-        psf = cupy.asnumpy(psf)
-    return psf
-
-
-def _psf(meshinfo, angles, diffraction_orders, *, focal_plane=False, use_gpu=True):
-    psf = np.zeros((4096, 4096), dtype=float)
-    if use_gpu and not HAS_CUPY:
-        log.info("cupy not installed or working, falling back to CPU")
-    # If cupy is available, cast to a cupy array
-    if HAS_CUPY and use_gpu:
-        psf = cupy.array(psf)
-    nx, ny = psf.shape
-    width_x = meshinfo["width"].value
-    width_y = meshinfo["width"].value
-    # x and y position grids
-    x = np.outer(np.ones(ny), np.arange(nx) + 0.5)
-    y = np.outer(np.arange(ny) + 0.5, np.ones(nx))
-    if HAS_CUPY and use_gpu:
-        x = cupy.array(x)
-        y = cupy.array(y)
-    area_not_mesh = 0.82  # fractional area not covered by the mesh
-    spacing = meshinfo["spacing_fp"] if focal_plane else meshinfo["spacing_e"]
+    psf = np.abs(np.fft.fft2(np.fft.fft2(psf_e) * np.fft.fft2(psf_f)))
+    # Center PSF at pixel (0,0)
+    psf = np.fft.fftshift(psf)
+    # Normalize by total number of pixels and always return a numpy array
+    return asarray(psf / (psf.shape[0] * psf.shape[1]))
+
+
+def _psf(meshinfo, angles, diffraction_orders, *, focal_plane=False):
+    ny, nx = 4096, 4096
+    width = meshinfo["width"].to_value("pixel")
+    spacing = (meshinfo["spacing_fp"] if focal_plane else meshinfo["spacing_e"]).to_value("pixel")
     mesh_ratio = (meshinfo["mesh_pitch"] / meshinfo["mesh_width"]).decompose().value
-    spacing_x = spacing * np.cos(angles)
-    spacing_y = spacing * np.sin(angles)
-    for order in diffraction_orders:
-        if order == 0:
-            continue
-        intensity = np.sinc(order / mesh_ratio) ** 2  # I_0
-        for dx, dy in zip(spacing_x.value, spacing_y.value, strict=True):
-            x_centered = x - (0.5 * nx + dx * order + 0.5)
-            y_centered = y - (0.5 * ny + dy * order + 0.5)
-            # NOTE: this step is the bottleneck and is VERY slow on a CPU
-            psf += np.exp(-width_x * x_centered * x_centered - width_y * y_centered * y_centered) * intensity
-    # Contribution from core
-    psf_core = np.exp(-width_x * (x - 0.5 * nx - 0.5) ** 2 - width_y * (y - 0.5 * ny - 0.5) ** 2)
-    return (1 - area_not_mesh) * psf / psf.sum() + area_not_mesh * psf_core / psf_core.sum()
+    area_not_mesh = 0.82
+    # 1-D coordinates and image center
+    x = np.arange(nx) + 0.5
+    y = np.arange(ny) + 0.5
+    cx0 = 0.5 * nx + 0.5
+    cy0 = 0.5 * ny + 0.5
+    # Per-angle pixel offsets
+    ang = np.asarray(angles.to_value(u.rad))
+    dxs = spacing * np.cos(ang)
+    dys = spacing * np.sin(ang)
+    orders = np.asarray(diffraction_orders)
+
+    @jax_jit
+    def _spikes_kernel(x, y, dxs, dys, orders, width, mesh_ratio, cx0, cy0):
+        width = np.asarray(width, dtype=x.dtype)
+        mesh_ratio = np.asarray(mesh_ratio, dtype=x.dtype)
+        cx0 = np.asarray(cx0, dtype=x.dtype)
+        cy0 = np.asarray(cy0, dtype=x.dtype)
```

Comment:
What is the point of the explicit type casting here?

Reply:
Removed.